File: src/backend/Application/Common/Models/Result.cs
namespace EvrenDev.Application.Common.Models;
public class Result<T>
{
    public bool Succeeded { get; private set; }
    public T Data { get; private set; }
    public string[] Errors { get; private set; }

    private Result(bool succeeded, T data = default!, string[] errors = null!)
    {
        Succeeded = succeeded;
        Data = data;
        Errors = errors ?? Array.Empty<string>();
    }

    public static Result<T> Success(T data)
    {
        return new Result<T>(true, data);
    }

    public static Result<T> Failure(params string[] errors)
    {
        return new Result<T>(false, errors: errors);
    }
}

File: src/frontend/src/primitives/Result.ts
import type { AppError } from './Error'

export class Result<T> {
  public readonly data?: T
  public readonly succeeded: boolean
  public readonly errors?: AppError

  private constructor(data?: T, errors?: AppError) {
    if (errors) {
      this.errors = errors
      this.succeeded = false
    } else {
      this.data = data
      this.succeeded = true
    }
  }

  public static success<T>(data: T): Result<T> {
    return new Result<T>(data)
  }

  public static failure<T>(errors: AppError): Result<T> {
    return new Result<T>(undefined, errors)
  }

  public static create<T>(data: T): Result<T> {
    return new Result<T>(data)
  }

  public get isFailure(): boolean {
    return !this.succeeded
  }
}
